File: Accupay.Utils/Extensions/DateExtensions.vb
Imports System.Runtime.CompilerServices

Module DateExtensions

    <Extension()>
    Public Function ToStringFormatOrNull(input As Date?, format As String) As String

        If input Is Nothing Then Return Nothing
        Return input.Value.ToString(format)

    End Function

    <Extension()>
    Public Function ToMinimumHourValue(input As Date) As Date

        Return New DateTime(input.Year, input.Month, input.Day, 0, 0, 0)

    End Function

    <Extension()>
    Public Function ToMaximumHourValue(input As Date) As Date

        Return New DateTime(input.Year, input.Month, input.Day, 23, 59, 59)

    End Function

    <Extension()>
    Public Function ToMinimumHourValue(input As Date?) As Date

        If input Is Nothing Then Return Nothing

        Return New DateTime(
            input.Value.Year, input.Value.Month, input.Value.Day, 0, 0, 0)

    End Function

    <Extension()>
    Public Function ToMaximumHourValue(input As Date?) As Date

        If input Is Nothing Then Return Nothing

        Return New DateTime(
            input.Value.Year, input.Value.Month, input.Value.Day, 23, 59, 59)

    End Function

End Module

File: AccuPay.Utilities/Extensions/DateExtensions.cs
using System;

namespace AccuPay.Utilities.Extensions
{
    public static class DateExtensions
    {
        public static string ToStringFormatOrNull(this DateTime? input, string format)
        {
            if (input == null) return null;
            return input.Value.ToString(format);
        }

        public static DateTime ToMinimumDateValue(this DateTime input)
        {
            return new DateTime(input.Year, input.Month, 1, 0, 0, 0);
        }

        public static DateTime ToMinimumHourValue(this DateTime input)
        {
            return new DateTime(input.Year, input.Month, input.Day, 0, 0, 0);
        }

        public static DateTime ToMaximumHourValue(this DateTime input)
        {
            return new DateTime(input.Year, input.Month, input.Day, 23, 59, 59);
        }

        public static DateTime? ToMinimumHourValue(this DateTime? input)
        {
            if (input == null) return null;

            return new DateTime(input.Value.Year, input.Value.Month, input.Value.Day, 0, 0, 0);
        }

        public static DateTime? ToMaximumHourValue(this DateTime? input)
        {
            if (input == null) return null;

            return new DateTime(input.Value.Year, input.Value.Month, input.Value.Day, 23, 59, 59);
        }

        public static DateTime StripSeconds(this DateTime input)
        {
            return new DateTime(input.Year, input.Month, input.Day, input.Hour, input.Minute, 0);
        }
    }
}